src/klein/app.py

```diff
@@ -329,12 +330,33 @@ def run(self, host, port, logFile=None):
 
         @param logFile: The file object to log to, by default C{sys.stdout}
         @type logFile: file object
+
+        @param proto: protocol string for endpoint, can be either "tcp" or "ssl"
+        @type proto: str
+
+        @param privateKey: filepath to private key for ssl
+        @type privateKey: str
+
+        @param certificate: filepath to certificate for ssl
+        @type certificate: str
+
+        @param dhParameters: filepath to dh parameters for ssl
+        @type dhParameters: str
         """
         if logFile is None:
             logFile = sys.stdout
 
         log.startLogging(logFile)
-        reactor.listenTCP(port, Site(self.resource()), interface=host)
+        server_string = "{}:{}:interface={}".format(proto, port, host)
+
+        if privateKey and certificate:
+            server_string += ":privateKey={}:certKey={}".format(privateKey,
+                                                                certificate)
+        if dhParameters:
+            server_string += ":dhParameters={}".format(dhParameters)
+
+        endpoint = endpoints.serverFromString(reactor, server_string)
+        endpoint.listen(Site(self.resource()))
```

Comment:
> why do you think it's better?

Because it allows users to use the full grammar of endpoints. For example, it's inconvenient in this form to use [txacme](https://txacme.readthedocs.io/en/latest/), because you end up having to smuggle it in by passing `proto="lets:/srv/www/certs:tcp", port="443"`. That *works*, but it reveals the weaknesses of the string formatting approach: it gives you enough rope to totally break something, but not enough to achieve something really powerful.

So if we're going to use `serverFromString`, we should just let the user pass whatever endpoint string they want. That gives them the full flexibility to ask for whatever they want. For example, they can pass TLS options other than the ones you allowed them, or use txacme to auto-provision their certs, or any number of other good things.